pocketutils/misc/klgists.py

```diff
@@ -0,0 +1,829 @@
+		"""Returns a new TomlData with its top set to items[1][2]..."""
+		items = key.split('.')
+		item = self.top
+		for i, s in enumerate(items):
+			if s not in item: raise MissingConfigEntry(
+				"{} is not in the TOML; failed at {}"
+				.format(key, '.'.join(items[:i+1]))
+			)
+			item = item[s]
+		return TomlData(item)
+
+	def items(self) -> ItemsView[str, object]:
+		return self.top.items()
+
+	def keys(self) -> KeysView[str]:
+		return self.top.keys()
+
+	def values(self) -> ValuesView[object]:
+		return self.top.values()
+
+	def nested_keys(self, separator='.') -> Iterator[str]:
+		for lst in self.nested_key_lists(self.top):
+			yield separator.join(lst)
+
+	def nested_key_lists(self, dictionary: Dict[str, object], prefix=None) -> Iterator[List[str]]:
+
+		prefix = prefix[:] if prefix else []
+
+		if isinstance(dictionary, dict):
+			for key, value in dictionary.items():
+
+				if isinstance(value, dict):
+					for result in self.nested_key_lists(value, [key] + prefix): yield result
+				else: yield prefix + [key]
+
+		else: yield dictionary
+
+
+def git_commit_hash(git_repo_dir: str='.') -> str:
+	"""Gets the hex of the most recent Git commit hash in git_repo_dir."""
+	p = Popen(['git', 'rev-parse', 'HEAD'], stdout=PIPE, cwd=git_repo_dir)
+	(out, err) = p.communicate()
+	exit_code = p.wait()
+	if exit_code != 0: raise ValueError("Got nonzero exit code {} from git rev-parse".format(exit_code))
+	return out.decode('utf-8').rstrip()
+
+@deprecated(reason="Use klgists.common.flexible_logger instead.")
+def init_logger(
+	log_path: Optional[str]=None,
+	format_str: str='%(asctime)s %(levelname)-8s: %(message)s',
+	to_std: bool=True,
+	child_logger_name: Optional[str]=None,
+	std_level = logging.INFO,
+	file_level = logging.DEBUG
+):
+	"""Initializes a logger that can write to a log file and/or stdout."""
+
+	if log_path is not None and not os.path.exists(os.path.dirname(log_path)):
+		os.mkdir(os.path.dirname(log_path))
+
+	if child_logger_name is None:
+		logger = logging.getLogger()
+	else:
+		logger = logging.getLogger(child_logger_name)
+	logger.setLevel(logging.NOTSET)
+
+	formatter = logging.Formatter(format_str)
+
+	if log_path is not None:
+		handler = logging.FileHandler(log_path, encoding='utf-8')
+		handler.setLevel(file_level)
+		handler.setFormatter(formatter)
+		logger.addHandler(handler)
+
+	if to_std:
+		handler = logging.StreamHandler()
+		handler.setLevel(std_level)
+		handler.setFormatter(formatter)
+		logger.addHandler(handler)
+
+import datetime
+def format_time(time: datetime) -> str:
+	"""Standard timestamp format. Ex: 2016-05-02_22_35_56."""
+	return time.strftime("%Y-%m-%d_%H-%M-%S")
+
+def timestamp() -> str:
+	"""Standard timestamp of time now. Ex: 2016-05-02_22_35_56."""
+	return format_time(datetime.datetime.now())
+
+def timestamp_path(path: str) -> str:
+	"""Standard way to label a file path with a timestamp."""
+	return "{}-{}".format(path, timestamp())
+
+
+def nice_time(n_ms: int) -> str:
```

Comment:
[UnitTools `ms_to_minsec`](https://github.com/dmyersturnbull/pocketutils/blob/main/pocketutils/tools/unit_tools.py#LL85C9-L85C21) is about the same.